packages/backend/src/apps/lettersg/actions/create-letter/index.ts

```diff
@@ -90,7 +116,26 @@ const action: IRawAction = {
       const rawResponse = await $.http.post('/v1/letters', payload)
       const response = responseSchema.parse(rawResponse.data)
 
-      $.setActionItem({ raw: response })
+      if (
+        !$.step.parameters.shouldGeneratePdf ||
+        !$.flow.hasFileProcessingActions
+      ) {
+        $.setActionItem({
+          raw: response,
+        })
+        return
+      }
+
+      const attachmentS3Key = await downloadAndStoreAttachmentInS3(
+        $,
+        response.publicId,
+      )
+      $.setActionItem({
+        raw: {
+          ...response,
+          attachment: attachmentS3Key,
+        },
+      })
     } catch (error) {
```

Comment:
nit: actually ah I was thinking of doing this instead, because I wanted to isolate the HTTP 400 error only to the `/v1/letters` call.

LMK what you think.

```
async function getCreateLetterRawResponse(
  $: IGlobalVariable,
) /* type inferred */ {
  try {
    const payload = requestSchema.parse($.step.parameters)
    // Don't parse here, because we have a special zod error message for
    // parameters parsing failures. That error message will be confusing if
    // the zod parsing failure is due to bad HTTP response.
    return await $.http.post('/v1/letters', payload)
  } catch (error) {
    if (error instanceof ZodError) {
      const firstError = fromZodError(error).details[0]
      throw new StepError(
        `${firstError.message} under set up action`,
        GenericSolution.ReconfigureInvalidField,
        $.step.position,
        $.app.name,
      )
    }

    // TODO (mal): check specific fields to return
    if (error instanceof HttpError && error.response.status === 400) {
      const lettersErrorData: LettersApiFieldErrorData = error.response.data
      if (lettersErrorData.message === 'Invalid letter params.') {
        throw new StepError(
          'Missing pair of field/value for letter template',
          'Click on set up action and check that you have entered all the fields and values in the letter parameters.',
          $.step.position,
          $.app.name,
        )
      }
    }

    // Bubble up
    throw error
  }
}

const action: IRawAction = {
  /* ... */
  async run($) {
    try {
      const rawResponse = await getCreateLetterRawResponse($)
      const response = responseSchema.parse(rawResponse)
      if (
        !$.step.parameters.shouldGeneratePdf ||
        !$.flow.hasFileProcessingActions
      ) {
        $.setActionItem({
          raw: response,
        })
        return
      }

      const attachmentS3Key = await downloadAndStoreAttachmentInS3(
        $,
        response.publicId,
      )
      $.setActionItem({
        raw: {
          ...response,
          attachment: attachmentS3Key,
        },
      })
    } catch (error) {
      if (!(error instanceof StepError)) {
        throw new StepError(
          `An error occurred: '${error.message}'`,
          'Please check that you have configured your step correctly',
          $.step.position,
          $.app.name,
        )
      }

      throw error
    }
  },
}
```

Reply:
thanks for pointing this out, will address this in another PR with the todo nits